### PR TITLE
PreRender

### PR DESCRIPTION
**Pull request by sastaachar**
PreRender

Requirements : 

1. preRender
2. showPreRender
3. hidePreRender
4. React components for preRender

# 1. PreRender

```
 preRender flow -> 
 ------------------------------------
   preRender                        ->   render()
     isPreRendered = true
  
  
Render flow -> 
------------------------------------
  render                ->     renderIframe
    isRendered = true      if ( isPreRendered ) 
                             insertIntoDOMForPreRender
                           else
                               insertIntoDOM
```

PreRender Flow

1. preRender function is called.
2. Set isPreRendered to true.
3. call render function

Render Flow

1. render function is called.
2. calls renderIframe.
3. If isPreRendered is true, call insertIntoDOMForPreRender.
    Otherwise, call insertIntoDOM.

## insertIntoDOMForPreRender
```
insertIntoDOMForPreRender( shouldShowByDefault = false ) -> 
      create preRenderWrapper
         <div> <iFrame> </div>
      if( shouldShowByDefault ) 
          showPreRender()
      else 
          hidePreRender()
      document.body.appendChild(preRenderWrapper);
```

This function is responsible for inserting the component into the DOM with optional visibility settings.
    
    
# 2. ShowPreRender
```
   if isPreRendered  -> we have rendered in this object
       preRenderWrapper 
           Remove inline styles of 'z-index', 'opacity', 'pointer-events'
       
       subscribeToEvents()
    else  -> preRender not called 
        check if preRender elements are there in dom
            if available set internal variables to point to them
            else preRender ( shouldShowByDefault = true ) 
```
             

1.  Create a preRenderWrapper containing a `<div>` wrapper for the child, `<iframe>` or in some cases a string (auth failure message) .
2. If shouldShowByDefault is true, call showPreRender(). Otherwise, call hidePreRender().
3. subscribe to the events
4. Append the preRenderWrapper to the document body.

# 3. HidePreRender
   ```
if isPreRendered  -> we have rendered in this object
       preRenderWrapper 
          setStyleProperties(this.preRenderWrapper, {
            opacity: '0',
            pointerEvents: 'none',
            zIndex: '-1000',
            position: 'absolute ',
            top: '0',
            left: '0',
        });
       unsubscribeToEvents()
    else  -> preRender not called 
         Do nothing
```

The HidePreRender module handles hiding the pre-rendered component.

If not isPreRendered:
      Do nothing.
If isPreRendered:
  1. Retrieve the preRenderWrapper.
  2. Apply styles to hide it (e.g., opacity: '0', pointerEvents: 'none', zIndex: '-1000').
  3. Unsubscribe from events.

    
   
# 4. React Components 
  
   Adding isPreRenderedComponent to factory
    
```
    if  isPreRenderedComponent 
        // Embed component like PreRenderedLiveboardEmbed 
        preRenderId is mandatory here for all Embed
        add preRender in useEffect instead of Render
        in useEffect clean up do not call destroy
         
    else 
       // Embed component like LiveboardEmbed    
       if preRenderId is passed in props 
           add showPreRender in useEffect 
           call hidePrePrender in useEffect clean up
```
For isPreRenderedComponent:

1. Embed components like PreRenderedLiveboardEmbed.
2. Ensure that preRenderId is mandatory for all embeds.
3. Add the preRender function in the useEffect instead of render.
4. In the useEffect cleanup, do not call destroy.

For non-isPreRenderedComponent:

1. Embed components like LiveboardEmbed.
2. If preRenderId is passed in props, add showPreRender in the useEffect. 
3. Call hidePreRender in the useEffect cleanup.

Demo : https://codesandbox.io/s/prerender-demo-mk4q6c

### CODE REVIEW

File: src/embed/ts-embed.ts [hunk 1]
```diff
@@ -592,12 +627,161 @@ export class TsEmbed {
...
+
+        const preRenderWrapper = document.createElement('div');
+        preRenderWrapper.id = preRenderIds.wrapper;
+        setStyleProperties(preRenderWrapper, { position: 'absolute', width: '100vw', height: '100vh' });
+
+        const preRenderShield = document.createElement('div');
+        preRenderShield.id = preRenderIds.shield;
+        setStyleProperties(preRenderShield, { position: 'absolute', width: '100%', height: '100%' });
+
+        child.id = preRenderIds.child;
```
Comment (pushkar100) [inline src/embed/ts-embed.ts:669]: Seems like a bug. Is this `child` element being implicitly created?

File: src/embed/ts-embed.ts [hunk 2]
```diff
@@ -592,12 +627,161 @@ export class TsEmbed {
...
+            stalePreRenderWrapper.remove();
+        }
+
+        const preRenderWrapper = document.createElement('div');
+        preRenderWrapper.id = preRenderIds.wrapper;
+        setStyleProperties(preRenderWrapper, { position: 'absolute', width: '100vw', height: '100vh' });
+
+        const preRenderShield = document.createElement('div');
+        preRenderShield.id = preRenderIds.shield;
+        setStyleProperties(preRenderShield, { position: 'absolute', width: '100%', height: '100%' });
```
Comment (pushkar100) [inline src/embed/ts-embed.ts:667]: Abstract away this frequently used logic.
```suggestion
const createUnqiueDivWithStyle = (id, styles) => {
    const div = document.createElement('div');
    div.setAttribute('id', id);
    setStyleProperties(div, styles);
    return div;
};

this.preRenderWrapper = createUnqiueDivWithStyle(preRenderIds.wrapper, { position: 'absolute', width: '100vw', height: '100vh' });
this.preRenderWrapper = createUnqiueDivWithStyle(preRenderIds.shield, { position: 'absolute', width: '100%', height: '100%' });
```

File: src/embed/ts-embed.ts [hunk 3]
```diff
@@ -829,6 +1013,16 @@ export class TsEmbed {
         return this;
     }
 
+    private isPreRendered: boolean;
+
+    /**
+     * Creates the preRender shell
+     */
+    public preRender(showPreRenderByDefault = false): TsEmbed {
```
Comment (pushkar100) [inline src/embed/ts-embed.ts:1021]: Feel that renaming `showPreRenderByDefault ` to something like `renderImmediately` conveys the option better

File: src/embed/ts-embed.ts [hunk 4]
```diff
@@ -592,12 +627,161 @@ export class TsEmbed {
...
+
+        document.body.appendChild(preRenderWrapper);
+
+        this.hidePreRender();
+    }
+
+    public hidePreRender(): void {
+        if (!this.isPreRenderAvailable()) {
+            // if the embed component is not preRendered , nothing to hide
+            console.log('No preRender found, not hiding ');
```
Comment (pushkar100) [inline src/embed/ts-embed.ts:724]: nit:- remove the log

File: src/embed/ts-embed.ts [hunk 5]
```diff
@@ -592,12 +627,161 @@ export class TsEmbed {
...
+            || document.getElementById(preRenderIds.shield);
+        this.preRenderChild = this.preRenderChild
+            || document.getElementById(preRenderIds.child);
+        this.iFrame = this.preRenderChild as HTMLIFrameElement;
+        return this.isPreRenderAvailable();
+    }
+
+    protected isPreRenderAvailable(): boolean {
+        return !!this.preRenderWrapper && !!this.preRenderShield
+            && !!this.preRenderChild;
```
Comment (pushkar100) [inline src/embed/ts-embed.ts:701]: nit:-
```suggestion
        return Boolean(this.preRenderWrapper && this.preRenderShield
            && this.preRenderChild);
```

File: src/embed/ts-embed.ts [hunk 6]
```diff
@@ -592,12 +627,161 @@ export class TsEmbed {
...
+
+    protected preRenderWrapper: HTMLElement;
+
+    protected preRenderShield: HTMLElement;
+
+    protected preRenderChild: HTMLElement;
+
+    protected connectPreRendered(): boolean {
+        const preRenderIds = this.getPreRenderIds();
+        this.preRenderWrapper = this.preRenderWrapper
```
Comment (pushkar100) [inline src/embed/ts-embed.ts:689]: Why will we need to re-read the wrapper/shield/child from the DOM once it has been assigned to the class property?

File: src/embed/ts-embed.ts [hunk 7]
```diff
@@ -555,7 +579,11 @@ export class TsEmbed {
             return getAuthPromise()
```
Comment (pushkar100) [inline src/embed/ts-embed.ts:579]: Branching into a different logic based on multiple `if`s on `this.isPreRendered` will lead to bloated, convoluted code.
Let us try to use a strategy pattern to split the logic between justr render and prerender. Ex:
```javascript
const renderStrategy = { success: () => { ... }, error: => { ... }};
const preRenderStrategy = { success: () => { ... }, error: => { ... }};
const setAuthPromiseStrategy = (isPreRendered) => {
    return isPreRendered ? preRenderStrategy : renderStrategy;
}
const authPromiseCallbacks = setAuthPromiseStrategy(this.isPreRendered);

return getAuthPromise(authPromiseCallbacks.success)?.then(authPromiseCallbacks.error);
```

**Review by pushkar100 — CHANGES_REQUESTED** [review]
Behaviour looks good. Need quite a bit of refactoring though

File: src/utils.ts [hunk 8]
```diff
@@ -248,3 +248,23 @@ export function getDOMNode(domSelector: DOMSelector): HTMLElement {
 }
 
 export const deepMerge = (target: any, source: any) => merge(target, source);
+
+export const setStyleProperties = (
+    element: HTMLElement,
```
[the remark this answers is not in the document]
Comment (sastaachar) [reply src/utils.ts:253]: There is no if here , i think you may have commented wrong line

[the remark this answers is not in the document]
Comment (pushkar100) [reply src/utils.ts:253]: Refer `if (!element || !styleProperties) return;`

File: src/embed/ts-embed.ts (hunk 1 above)
Comment (sastaachar) [reply src/embed/ts-embed.ts:668]: no its passed as parameter for the function , its the child that we are going to render

File: src/embed/ts-embed.ts (hunk 3 above)
Comment (sastaachar) [reply src/embed/ts-embed.ts:1020]: we render immediately all the time, the difference here is that we hide it by default , thats y i had named it show by default


render -> hide (normal flow)
render -> show (showPreRenderByDefault flow)

File: src/embed/ts-embed.ts (hunk 4 above)
Comment (sastaachar) [reply src/embed/ts-embed.ts:723]: yes i will be removing all the logs and cleaning , PS : it does not give any lint error (it should)

File: src/embed/ts-embed.ts (hunk 6 above)
Comment (sastaachar) [reply src/embed/ts-embed.ts:688]: this is for the case where another object tries to connect to our preRender

File: src/embed/ts-embed.ts (hunk 7 above)
Comment (sastaachar) [reply src/embed/ts-embed.ts:579]: 90% of the code is similar for both strategy , I'll move this logic to another function which will make it more clear as well and branch their based on the strategy

File: src/embed/ts-embed.ts (hunk 3 above)
Comment (pushkar100) [reply src/embed/ts-embed.ts:1024]: `showImmediately` makes sense to me. The `bydefault` is what is confusing

**Review by pushkar100 — APPROVED** [review]
LGTM. Let's organise a bug hunt

**Comment by pushkar100** [comment]
@Nirmay15 Can you please review this too?

File: src/react/index.tsx [hunk 9]
```diff
@@ -13,63 +13,79 @@ import { EmbedProps, getViewPropsAndListeners } from './util';
...
+    * Use PreRender to render the embed in the background and then
+    * show or hide the rendered embed using showPreRender or hidePreRender respectively.
+    *
+    * @example
+    * ```js
+    * const embed = new LiveboardEmbed('#embed', {
+    *   ... // other liveboard view config
+    *   preRenderId: "preRenderId-123"
+    * });
+    * embed.showPreRender();
```
Comment (ashubham) [inline src/react/index.tsx:82]: Lets have a React example here.

File: src/types.ts [hunk 10]
```diff
@@ -791,6 +792,41 @@ export interface ViewConfig {
...
+     * @example
+     * // Enable tracking PreRender size in the configuration
+     * const config = {
+     *   trackPreRenderSize: true,
+     * };
+     *
+     * // Instantiate an object with the configuration
+     * const myComponent = new MyComponent(config);
+     */
+    trackPreRenderSize?: boolean;
```
Comment (ashubham) [inline src/types.ts:829]: Lets always have default values as `false`. Change this to `doNotTrackPrerenderSize`

File: src/react/index.tsx [hunk 11]
```diff
@@ -13,63 +13,79 @@ import { EmbedProps, getViewPropsAndListeners } from './util';
 
 const componentFactory = <T extends typeof TsEmbed, U extends EmbedProps, V extends ViewConfig>(
     EmbedConstructor: T,
+    isPreRenderedComponent = false,
```
Comment (ashubham) [inline src/react/index.tsx:16]: Lets put comments to what this does.

Also, I see multiple f else conditions inside, is it possible to structure this to be more composed ?

**Pull request merged**